#  NGS Resources

## Files for NGS  alignment  workflow
* BBMap adapter: Adapter used with  BBMap. Downloaded [BBMap adapters.fa](https://github.com/BioInfoTools/BBMap/blob/master/resources/adapters.fa).
* HISAT2 index: Used with HISAT2 during alignment to map  reads to reference genome. Download [HISAT2 genome/](https://daehwankimlab.github.io/hisat2/download/).
* RefSeq reference genome: NCBI RefSeq h38 reference genome. Download [NCBI hgg38  refseq bed](https://www.ncbi.nlm.nih.gov/refseq/)
* Chromosome size: IGV chrome size downloaded from [IGV hg38.chrom.sizes](https://github.com/igvteam/igv/blob/master/genomes/sizes/hg38.chrom.sizes).
* read correction count: readcountcorrectBG.py program available on [mapandvischainonfiji github](https://github.com/maallen3/mapandvischainonfiji). Use rcc to reduce sequencing bias based on depth for visualization in IGV.
* iggvtools: Details on [igvtools toTDF](https://github.com/philres/IGV/blob/master/docs/igvtools_readme.txt). 
